src/Interpreters/ComparisonTupleEliminationVisitor.cpp

```diff
@@ -0,0 +1,108 @@
+#include <Interpreters/ComparisonTupleEliminationVisitor.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+
+
+ASTs splitTuple(const ASTPtr & node)
+{
+    if (const auto * func = node->as<ASTFunction>(); func && func->name == "tuple")
+        return func->arguments->children;
+
+    if (const auto * literal = node->as<ASTLiteral>(); literal && literal->value.getType() == Field::Types::Tuple)
+    {
+        ASTs result;
+        const auto & tuple = literal->value.get<const Tuple &>();
+        for (const auto & child : tuple)
+            result.emplace_back(std::make_shared<ASTLiteral>(child));
+        return result;
+    }
+
+    return {};
+}
+
+
+ASTPtr concatWithAnd(const ASTs & nodes)
+{
+    if (nodes.empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot concat empty list of nodes");
+
+    if (nodes.size() == 1)
+        return nodes[0];
+
+    auto result = makeASTFunction("and");
+    result->arguments->children = nodes;
+    return result;
+}
+
+void trySplitTupleComparsion(ASTPtr & expression)
+{
+    if (!expression)
+        return;
+
+    auto * func = expression->as<ASTFunction>();
+    if (!func)
+        return;
+
+    if (func->name == "and" || func->name == "or" || func->name == "not" || func->name == "tuple")
+    {
+        for (auto & child : func->arguments->children)
+        {
+            trySplitTupleComparsion(child);
+        }
+    }
+
+    if (func->name == "equals" || func->name == "notEquals")
+    {
+        if (func->arguments->children.size() != 2)
+            return;
+
+        auto lhs = splitTuple(func->arguments->children[0]);
+        auto rhs = splitTuple(func->arguments->children[1]);
+        if (lhs.size() != rhs.size() || lhs.empty() || rhs.empty())
+            return;
+
+        ASTs new_args;
+        for (size_t i = 0; i < lhs.size(); ++i)
+        {
+            trySplitTupleComparsion(lhs[i]);
```

Comment:
Not sure why we need to dive into lhs and rhs ? If we already understand that lhs and rhs are tuples ?

Reply:
I wanted to cover case `((a, b), 1000) = ((c, 10), d)` to have `(a = c AND b = 10) AND 1000 = d`.

But according to my test it's still expanded to `(((a, b) = (c, 10)) AND (1000 = d))` 😅

Reply:
Do we actually want do expand nested tuples?